twill-core/src/main/java/org/apache/twill/internal/AbstractTwillService.java

```diff
@@ -216,11 +217,52 @@ protected final void shutDown() throws Exception {
     return zkClient.setData(liveNodePath, serializeLiveNode());
   }
 
+  /**
+   * Creates the live node for the service. If the node already exists, it will be deleted before creation.
+   *
+   * @return A {@link OperationFuture} that will be completed when the creation is done.
+   */
   private OperationFuture<String> createLiveNode() {
-    String liveNodePath = getLiveNodePath();
+    final String liveNodePath = getLiveNodePath();
     LOG.info("Create live node {}{}", zkClient.getConnectString(), liveNodePath);
-    return ZKOperations.ignoreError(zkClient.create(liveNodePath, serializeLiveNode(), CreateMode.EPHEMERAL),
-                                    KeeperException.NodeExistsException.class, liveNodePath);
+    final SettableOperationFuture<String> resultFuture = SettableOperationFuture.create(liveNodePath,
+                                                                                        Threads.SAME_THREAD_EXECUTOR);
+    OperationFuture<String> createFuture = zkClient.create(liveNodePath, serializeLiveNode(), CreateMode.EPHEMERAL);
+    Futures.addCallback(createFuture, new FutureCallback<String>() {
+      final FutureCallback<String> thisCallback = this;
+
+      @Override
+      public void onSuccess(String result) {
+        LOG.info("Live node created {}{}", zkClient.getConnectString(), liveNodePath);
+        resultFuture.set(result);
+      }
+
+      @Override
+      public void onFailure(final Throwable createFailure) {
+        if (!(createFailure instanceof KeeperException.NodeExistsException)) {
+          resultFuture.setException(createFailure);
+        }
+
+        // If the node already exists, it is due to previous run attempt that left an ephemeral node.
```

Comment:
Ephemeral node won't go away immediate if the process die. It will stay there till ZK session timeout, which is typically multiple seconds. In the meantime, the next AM process may already be started by YARN, hence the new AM process will see the ephemeral node.

Reply:
got it